src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/Error/SqlExtendedQueryTagErrorStoreV4.cs

```diff
@@ -0,0 +1,109 @@
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            ConnectionWrapperFactory = sqlConnectionWrapperFactory;
+            Logger = logger;
+        }
+
+        public override SchemaVersion Version => SchemaVersion.V4;
+
+        protected SqlConnectionWrapperFactory ConnectionWrapperFactory { get; }
+
+        protected ILogger Logger { get; }
+
+        public override async Task<int> AddExtendedQueryTagErrorAsync(
+            int tagKey,
+            int errorCode,
+            long watermark,
+            DateTime createdTime,
+            CancellationToken cancellationToken = default)
+        {
+            using SqlConnectionWrapper sqlConnectionWrapper = await ConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken);
+            using SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand();
+            VLatest.AddExtendedQueryTagError.PopulateCommand(
+                sqlCommandWrapper,
+                tagKey,
+                errorCode,
+                watermark,
+                createdTime);
+            try
+            {
+                return (int)await sqlCommandWrapper.ExecuteScalarAsync(cancellationToken);
+            }
+            catch (SqlException e)
+            {
+                if (e.Number == SqlErrorCodes.NotFound)
+                    throw new ExtendedQueryTagNotFoundException("Attempted to add error on non existing query tag.");
+
+                throw new DataStoreException(e);
+            }
+        }
+
+        public override async Task<IReadOnlyList<ExtendedQueryTagError>> GetExtendedQueryTagErrorsAsync(string tagPath, CancellationToken cancellationToken = default)
+        {
+            List<ExtendedQueryTagError> results = new List<ExtendedQueryTagError>();
+
+            using SqlConnectionWrapper sqlConnectionWrapper = await ConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken);
+            using SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand();
+
+            VLatest.GetExtendedQueryTagErrors.PopulateCommand(sqlCommandWrapper, tagPath);
+
+            using SqlDataReader reader = await sqlCommandWrapper.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken);
+            while (await reader.ReadAsync(cancellationToken))
+            {
+                (int tagkey, int errorCode, DateTime createdTime, string studyInstanceUid, string seriesInstanceUid, string sopInstanceUid) = reader.ReadRow(
+                    VLatest.ExtendedQueryTagError.TagKey,
+                    VLatest.ExtendedQueryTagError.ErrorCode,
+                    VLatest.ExtendedQueryTagError.CreatedTime,
+                    VLatest.Instance.StudyInstanceUid,
+                    VLatest.Instance.SeriesInstanceUid,
+                    VLatest.Instance.SopInstanceUid);
+
+                //TODO: build the error message here
+                string errorMessage = "error: " + errorCode;
```

Comment:
As ErrorCode concept is defined in Core project, maybe it's a good idea to build error message in Core.

Reply:
Will address in a separate PR